lib/krane/deploy_task.rb

```diff
@@ -1,12 +1,394 @@
 # frozen_string_literal: true
+require 'yaml'
+require 'shellwords'
+require 'tempfile'
+require 'fileutils'
 
-require 'krane/deprecated_deploy_task'
+require 'krane/common'
+require 'krane/concurrency'
+require 'krane/resource_cache'
+require 'krane/kubernetes_resource'
+%w(
+  custom_resource
+  cloudsql
+  config_map
+  deployment
+  ingress
+  persistent_volume_claim
+  pod
+  network_policy
+  service
+  pod_template
+  pod_disruption_budget
+  replica_set
+  service_account
+  daemon_set
+  resource_quota
+  stateful_set
+  cron_job
+  job
+  custom_resource_definition
+  horizontal_pod_autoscaler
+  secret
+).each do |subresource|
+  require "krane/kubernetes_resource/#{subresource}"
+end
+require 'krane/resource_watcher'
+require 'krane/kubectl'
+require 'krane/kubeclient_builder'
+require 'krane/ejson_secret_provisioner'
+require 'krane/renderer'
+require 'krane/cluster_resource_discovery'
+require 'krane/template_sets'
+require 'krane/deploy_task_config_validator'
+require 'krane/resource_deployer'
+require 'krane/concerns/template_reporting'
 
 module Krane
-  class DeployTask < Krane::DeprecatedDeployTask
-    def initialize(**args)
-      raise "Use Krane::DeployGlobalTask to deploy global resources" if args[:allow_globals]
-      super(args.merge(allow_globals: false))
+  # Ship resources to a namespace
+  class DeployTask
+    extend Krane::StatsD::MeasureMethods
+    include Krane::TemplateReporting
+
+    PROTECTED_NAMESPACES = %w(
+      default
+      kube-system
+      kube-public
+    )
+    # Things removed from default prune whitelist at https://github.com/kubernetes/kubernetes/blob/0dff56b4d88ec7551084bf89028dbeebf569620e/pkg/kubectl/cmd/apply.go#L411:
+    # core/v1/Namespace -- not namespaced
+    # core/v1/PersistentVolume -- not namespaced
+    # core/v1/Endpoints -- managed by services
+    # core/v1/PersistentVolumeClaim -- would delete data
+    # core/v1/ReplicationController -- superseded by deployments/replicasets
+
+    def predeploy_sequence
+      before_crs = %w(
+        ResourceQuota
+        NetworkPolicy
+      )
+      after_crs = %w(
+        ConfigMap
+        PersistentVolumeClaim
+        ServiceAccount
+        Role
+        RoleBinding
+        Secret
+        Pod
+      )
+
+      before_crs + cluster_resource_discoverer.crds.select(&:predeployed?).map(&:kind) + after_crs
+    end
+
+    def prune_whitelist
```

Comment:
This is going to conflict with https://github.com/Shopify/kubernetes-deploy/pull/616.

Reply:
I'm rebasing it and will update this PR soon.